Simbody/src/CableSpan.cpp

```diff
@@ -0,0 +1,3028 @@
+    }
+
+    // Mutable CableSpanData::Vel cache access.
+    CableSpanData::Vel& updDataVel(const State& state) const
+    {
+        return Value<CableSpanData::Vel>::updDowncast(
+            getSubsystem().updCacheEntry(state, m_indexDataVel));
+    }
+
+    //------------------------------------------------------------------------------
+    //  Computing cached data
+    //------------------------------------------------------------------------------
+
+    const MatrixWorkspace& calcDataInst(const State& state) const;
+
+    const CableSpanData::Pos& calcDataPos(const State& s) const
+    {
+        CableSpanData::Pos& dataPos = updDataPos(s);
+
+        dataPos.originPoint_G      = calcOriginPointInGround(s);
+        dataPos.terminationPoint_G = calcTerminationPointInGround(s);
+
+        // Helper for computing the total cable length.
+        auto calcTotalCableLength =
+            [&](const std::vector<LineSegment>& lines) -> Real
+        {
+            Real totalCableLength = 0.;
+            // Add length of each line segment.
+            for (const LineSegment& line : lines) {
+                totalCableLength += line.length;
+            }
+            // Add length of each curve segment.
+            forEachActiveCurveSegment(
+                s,
+                [&](const CurveSegment& curve)
+                {
+                    // Update the total cable length.
+                    totalCableLength += curve.getDataInst(s).arcLength;
+                });
+            return totalCableLength;
+        };
+
+        // Computing the CableSpan's path is done iteratively by computing
+        // corrections for the CurveSegments until the pathErrorVector is small
+        // enough.
+        dataPos.loopIter = 0;
+        while (true) {
+            // Compute all data required for updating the CableSpanData::Pos
+            // cache.
+            const MatrixWorkspace& data = calcDataInst(s);
+
+            // Stop iterating if:
+            // - No obstacles in contact with path: Path is straight line.
+            // - Converged: No significant correction.
+            // - Not converged: Max iterations has been reached.
+            if (data.nObstaclesInContact == 0 ||
+                data.pathCorrection.normInf() < Eps ||
+                dataPos.loopIter >= getParameters().solverMaxIterations) {
+                // Update cache entry and stop solver.
+                dataPos.smoothness  = data.maxPathError;
+                dataPos.cableLength = calcTotalCableLength(data.lineSegments);
+                dataPos.originTangent_G = data.lineSegments.front().direction;
+                dataPos.terminationTangent_G =
+                    data.lineSegments.back().direction;
+                break;
+            }
+
+            // Apply the corrections to each CurveSegment to reduce the path
+            // error.
+            const NaturalGeodesicCorrection* corrIt =
```

Comment:
If this function can't return a `nullptr` then it should return a reference? If it can return a `nullptr` then calling code should nullcheck

Reply:
no longer relevant after fixing https://github.com/simbody/simbody/pull/791#discussion_r1668456950